File: ios/Sasquatsh/Views/Toolbox/PickerSoundEngine.swift
import AVFoundation

/// Bell-based sound engine for the First Player Picker.
/// All sounds derived from bell/chime tones with musical pitch variation.
class PickerSoundEngine {
    static let shared = PickerSoundEngine()

    private let engine = AVAudioEngine()
    private let playerNode = AVAudioPlayerNode()
    private let format: AVAudioFormat

    // C major scale frequencies (C4 through C5)
    private let scale: [Double] = [
        261.63, // C4
        293.66, // D4
        329.63, // E4
        349.23, // F4
        392.00, // G4
        440.00, // A4
        493.88, // B4
        523.25, // C5
    ]

    // Pre-generated buffers
    private var tapBuffers: [AVAudioPCMBuffer] = []   // one per scale note
    private let transitionBuffer: AVAudioPCMBuffer     // reverse bell swell
    private var tickBuffers: [AVAudioPCMBuffer] = []   // damped bells, rising pitch
    private let chimeBuffer: AVAudioPCMBuffer           // full bell

    private init() {
        let fmt = AVAudioFormat(standardFormatWithSampleRate: 44100, channels: 1)!
        format = fmt

        let scaleFreqs = scale

        // Tap: short muted bell per scale note
        tapBuffers = scaleFreqs.map { freq in
            PickerSoundEngine.makeMutedBell(format: fmt, frequency: freq, duration: 0.08, volume: 0.4)
        }

        // Transition: soft reverse bell swell
        transitionBuffer = PickerSoundEngine.makeReverseBell(format: fmt)

        // Countdown ticks: damped bells ascending the scale
        tickBuffers = scaleFreqs.map { freq in
            PickerSoundEngine.makeDampedBell(format: fmt, frequency: freq, duration: 0.06, volume: 0.35)
        }

        // Final: full resonant bell chime
        chimeBuffer = PickerSoundEngine.makeFullBell(format: fmt)

        engine.attach(playerNode)
        engine.connect(playerNode, to: engine.mainMixerNode, format: format)

        do {
            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default, options: .mixWithOthers)
            try engine.start()
        } catch {
            print("PickerSoundEngine: \(error)")
        }
    }

    private func play(_ buffer: AVAudioPCMBuffer) {
        if !engine.isRunning { try? engine.start() }
        playerNode.scheduleBuffer(buffer, at: nil, options: [], completionHandler: nil)
        if !playerNode.isPlaying { playerNode.play() }
    }

    // MARK: - Public API

    /// Short muted bell for finger placement. Pitch mapped to C major scale by finger index.
    func playTap(fingerIndex: Int = 0) {
        let idx = fingerIndex % tapBuffers.count
        playerNode.volume = Float(0.35 + Double.random(in: -0.03...0.03))
        play(tapBuffers[idx])
    }

    /// Soft reverse bell swell when entering countdown.
    func playTransition() {
        playerNode.volume = 0.4
        play(transitionBuffer)
    }

    /// Damped bell tick for countdown. Progress (0–1) selects ascending pitch.
    func playCountdownTick(progress: Double = 0) {
        let idx = min(Int(progress * Double(tickBuffers.count - 1)), tickBuffers.count - 1)
        playerNode.volume = Float(0.3 + progress * 0.15)
        play(tickBuffers[max(0, idx)])
    }

    /// Full resonant bell chime for final selection.
    func playSelectChime() {
        playerNode.volume = 0.6
        play(chimeBuffer)
    }

    // MARK: - Bell Synthesis

    /// Short muted bell: fundamental + 2nd partial, fast damped envelope (~80ms)
    private static func makeMutedBell(format: AVAudioFormat, frequency: Double, duration: Double, volume: Double) -> AVAudioPCMBuffer {
        let sr = format.sampleRate
        let count = AVAudioFrameCount(sr * duration)
        let buf = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: count)!
        buf.frameLength = count
        let out = buf.floatChannelData![0]

        for i in 0..<Int(count) {
            let t = Double(i) / sr
            // Soft attack (no click): use sine-shaped onset ~3ms
            let attack = min(t / 0.003, 1.0)
            // Fast exponential decay
            let decay = exp(-t * 35)
            let env = attack * decay

            let f1 = sin(2.0 * .pi * frequency * t)
            let f2 = sin(2.0 * .pi * frequency * 2.2 * t) * 0.3 // inharmonic partial (bell-like)
            out[i] = Float((f1 + f2) * env * volume)
        }
        return buf
    }

    /// Reverse bell swell: amplitude rises then cuts, ~350ms
    private static func makeReverseBell(format: AVAudioFormat) -> AVAudioPCMBuffer {
        let sr = format.sampleRate
        let duration = 0.35
        let count = AVAudioFrameCount(sr * duration)
        let buf = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: count)!
        buf.frameLength = count
        let out = buf.floatChannelData![0]

        for i in 0..<Int(count) {
            let t = Double(i) / sr
            let norm = t / duration

            // Reverse envelope: slow rise, soft peak, gentle fall at end
            let env = pow(norm, 2.0) * (1.0 - pow(max(0, norm - 0.85) / 0.15, 2.0)) * 0.45

            // Bell tone cluster (G4 + B4 + D5) — G major chord shimmer
            let f1 = sin(2.0 * .pi * 392.0 * t)         // G4
            let f2 = sin(2.0 * .pi * 493.88 * t) * 0.6  // B4
            let f3 = sin(2.0 * .pi * 587.33 * t) * 0.3  // D5
            out[i] = Float((f1 + f2 + f3) * env)
        }
        return buf
    }

    /// Damped bell tick: like a muted bell strike, very short (~60ms)
    private static func makeDampedBell(format: AVAudioFormat, frequency: Double, duration: Double, volume: Double) -> AVAudioPCMBuffer {
        let sr = format.sampleRate
        let count = AVAudioFrameCount(sr * duration)
        let buf = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: count)!
        buf.frameLength = count
        let out = buf.floatChannelData![0]

        for i in 0..<Int(count) {
            let t = Double(i) / sr
            // Soft onset + fast damp
            let attack = min(t / 0.002, 1.0)
            let decay = exp(-t * 45)
            let env = attack * decay

            let f1 = sin(2.0 * .pi * frequency * t)
            let f2 = sin(2.0 * .pi * frequency * 3.0 * t) * 0.15 // bell partial
            out[i] = Float((f1 + f2) * env * volume)
        }
        return buf
    }

    /// Full resonant bell: rich harmonics, slow decay (~500ms)
    private static func makeFullBell(format: AVAudioFormat) -> AVAudioPCMBuffer {
        let sr = format.sampleRate
        let duration = 0.5
        let count = AVAudioFrameCount(sr * duration)
        let buf = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: count)!
        buf.frameLength = count
        let out = buf.floatChannelData![0]

        // C5 bell with inharmonic partials (characteristic bell sound)
        let partials: [(freq: Double, amp: Double, decay: Double)] = [
            (523.25, 1.0,  4.0),   // C5 fundamental
            (659.25, 0.6,  5.0),   // E5 (major third)
            (1046.5, 0.4,  6.0),   // C6 octave
            (1174.7, 0.25, 8.0),   // ~D6 (inharmonic, bell character)
            (1568.0, 0.15, 10.0),  // G6 shimmer
            (2093.0, 0.08, 12.0),  // C7 sparkle
        ]

        for i in 0..<Int(count) {
            let t = Double(i) / sr
            let attack = min(t / 0.005, 1.0) // 5ms soft onset
            var sample = 0.0

            for p in partials {
                sample += sin(2.0 * .pi * p.freq * t) * p.amp * exp(-t * p.decay)
            }

            out[i] = Float(sample * attack * 0.35)
        }
        return buf
    }
}

File: android/app/src/main/java/com/sasquatsh/app/views/toolbox/PickerSoundEngine.kt
package com.sasquatsh.app.views.toolbox

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import kotlin.math.PI
import kotlin.math.exp
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin

/**
 * Bell-based sound engine for the First Player Picker.
 *
 * Pre-generates PCM buffers for tap, transition, countdown, and select-chime
 * sounds, then plays them via [AudioTrack] on a background thread.
 *
 * Thread-safe: every public method can be called from any thread.
 */
object PickerSoundEngine {

    // ── constants ───────────────────────────────────────────────────────
    private const val SAMPLE_RATE = 44100
    private const val TWO_PI = 2.0 * PI

    /** C-major scale frequencies used for tap and countdown pitch mapping. */
    private val C_MAJOR_SCALE = doubleArrayOf(
        261.63, 293.66, 329.63, 349.23,
        392.00, 440.00, 493.88, 523.25
    )

    // ── pre-generated buffers ───────────────────────────────────────────
    private val tapBuffers: Array<ShortArray>
    private val transitionBuffer: ShortArray
    private val countdownBuffers: Array<ShortArray>
    private val selectChimeBuffer: ShortArray

    // ── audio format shared by all tracks ───────────────────────────────
    private val audioAttributes: AudioAttributes = AudioAttributes.Builder()
        .setUsage(AudioAttributes.USAGE_GAME)
        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
        .build()

    private val audioFormat: AudioFormat = AudioFormat.Builder()
        .setSampleRate(SAMPLE_RATE)
        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
        .build()

    // ── playback lock (serialises writes to AudioTrack) ─────────────────
    private val playbackLock = Any()

    init {
        // Pre-generate all buffers at load time so playback is instant.
        tapBuffers = Array(C_MAJOR_SCALE.size) { i -> generateTapBuffer(i) }

        transitionBuffer = generateTransitionBuffer()

        // 11 steps (progress 0.0 .. 1.0 in 0.1 increments) is plenty;
        // at runtime we pick the nearest bucket.
        countdownBuffers = Array(11) { i -> generateCountdownBuffer(i / 10.0) }

        selectChimeBuffer = generateSelectChimeBuffer()
    }

    // ══════════════════════════════════════════════════════════════════════
    //  PUBLIC API
    // ══════════════════════════════════════════════════════════════════════

    /** Short muted bell mapped to a C-major scale note. */
    fun playTap(fingerIndex: Int) {
        val idx = fingerIndex.coerceIn(0, C_MAJOR_SCALE.lastIndex)
        playBuffer(tapBuffers[idx])
    }

    /** Reverse bell swell – G major chord shimmer. */
    fun playTransition() {
        playBuffer(transitionBuffer)
    }

    /** Damped bell tick whose pitch rises with [progress] (0..1). */
    fun playCountdownTick(progress: Double) {
        val p = progress.coerceIn(0.0, 1.0)
        val idx = (p * 10).toInt().coerceIn(0, 10)
        playBuffer(countdownBuffers[idx])
    }

    /** Full resonant bell with rich harmonics. */
    fun playSelectChime() {
        playBuffer(selectChimeBuffer)
    }

    // ══════════════════════════════════════════════════════════════════════
    //  BUFFER GENERATION
    // ══════════════════════════════════════════════════════════════════════

    /** ~80 ms muted bell for finger tap. */
    private fun generateTapBuffer(scaleIndex: Int): ShortArray {
        val durationSec = 0.08
        val numSamples = (SAMPLE_RATE * durationSec).toInt()
        val buffer = ShortArray(numSamples)

        val freq = C_MAJOR_SCALE[scaleIndex]
        val partialFreq = freq * 2.2
        val volume = 0.4
        val attackSec = 0.003

        for (i in 0 until numSamples) {
            val t = i.toDouble() / SAMPLE_RATE

            // Envelope: soft 3 ms attack, fast exponential decay
            val attack = if (t < attackSec) t / attackSec else 1.0
            val decay = exp(-t * 35.0)
            val envelope = attack * decay

            // Fundamental + inharmonic partial
            val sample = sin(TWO_PI * freq * t) +
                    0.3 * sin(TWO_PI * partialFreq * t)

            // Slight random volume variation is baked per-buffer (deterministic
            // per finger index keeps it lightweight).
            val variation = 1.0 + (scaleIndex % 3 - 1) * 0.03
            buffer[i] = toShort(sample * envelope * volume * variation)
        }
        return buffer
    }

    /** ~350 ms reverse bell swell – G major chord shimmer. */
    private fun generateTransitionBuffer(): ShortArray {
        val durationSec = 0.35
        val numSamples = (SAMPLE_RATE * durationSec).toInt()
        val buffer = ShortArray(numSamples)

        val volume = 0.45

        // G major chord: G4, B4, D5
        val freqs = doubleArrayOf(392.0, 493.88, 587.33)
        val amps = doubleArrayOf(1.0, 0.6, 0.3)

        for (i in 0 until numSamples) {
            val t = i.toDouble() / SAMPLE_RATE
            val norm = t / durationSec // 0..1

            // Envelope: quadratic rise then gentle fall at the tail
            val rise = norm * norm
            val fall = if (norm > 0.85) 1.0 - ((norm - 0.85) / 0.15) else 1.0
            val envelope = rise * fall

            var sample = 0.0
            for (j in freqs.indices) {
                sample += amps[j] * sin(TWO_PI * freqs[j] * t)
            }

            buffer[i] = toShort(sample * envelope * volume)
        }
        return buffer
    }

    /** ~60 ms damped bell tick whose pitch ascends with [progress]. */
    private fun generateCountdownBuffer(progress: Double): ShortArray {
        val durationSec = 0.06
        val numSamples = (SAMPLE_RATE * durationSec).toInt()
        val buffer = ShortArray(numSamples)

        // Map progress to scale
        val scaleIdx = (progress * (C_MAJOR_SCALE.size - 1)).toInt()
            .coerceIn(0, C_MAJOR_SCALE.lastIndex)
        val freq = C_MAJOR_SCALE[scaleIdx]
        val partialFreq = freq * 3.0
        val volume = 0.3 + progress * 0.15
        val attackSec = 0.002

        for (i in 0 until numSamples) {
            val t = i.toDouble() / SAMPLE_RATE

            val attack = if (t < attackSec) t / attackSec else 1.0
            val decay = exp(-t * 45.0)
            val envelope = attack * decay

            val sample = sin(TWO_PI * freq * t) +
                    0.15 * sin(TWO_PI * partialFreq * t)

            buffer[i] = toShort(sample * envelope * volume)
        }
        return buffer
    }

    /** ~500 ms full resonant bell with rich harmonics. */
    private fun generateSelectChimeBuffer(): ShortArray {
        val durationSec = 0.5
        val numSamples = (SAMPLE_RATE * durationSec).toInt()
        val buffer = ShortArray(numSamples)

        val volume = 0.35
        val onsetSec = 0.005

        // Partials: frequency, amplitude, decay rate
        val partials = arrayOf(
            Triple(523.25, 1.00, 4.0),   // C5
            Triple(659.25, 0.60, 5.0),   // E5
            Triple(1046.50, 0.40, 6.0),  // C6
            Triple(1174.70, 0.25, 8.0),  // ~D6
            Triple(1568.00, 0.15, 10.0), // G6
            Triple(2093.00, 0.08, 12.0)  // C7
        )

        for (i in 0 until numSamples) {
            val t = i.toDouble() / SAMPLE_RATE

            // 5 ms soft onset
            val onset = if (t < onsetSec) t / onsetSec else 1.0

            var sample = 0.0
            for ((freq, amp, decayRate) in partials) {
                sample += amp * exp(-t * decayRate) * sin(TWO_PI * freq * t)
            }

            buffer[i] = toShort(sample * onset * volume)
        }
        return buffer
    }

    // ══════════════════════════════════════════════════════════════════════
    //  PLAYBACK
    // ══════════════════════════════════════════════════════════════════════

    /**
     * Plays the given PCM [buffer] on a fire-and-forget background thread.
     *
     * A short-lived [AudioTrack] in static mode is used so that each sound
     * can overlap with others without waiting for the previous one to finish.
     */
    private fun playBuffer(buffer: ShortArray) {
        Thread({
            try {
                val bufferBytes = buffer.size * 2 // 16-bit = 2 bytes per sample
                val track = AudioTrack.Builder()
                    .setAudioAttributes(audioAttributes)
                    .setAudioFormat(audioFormat)
                    .setBufferSizeInBytes(bufferBytes)
                    .setTransferMode(AudioTrack.MODE_STATIC)
                    .build()

                track.write(buffer, 0, buffer.size)
                track.play()

                // Wait for playback to finish then release resources.
                val durationMs = (buffer.size * 1000L) / SAMPLE_RATE + 50
                Thread.sleep(durationMs)

                track.stop()
                track.release()
            } catch (_: Exception) {
                // Silently swallow audio errors – sound is non-critical UI feedback.
            }
        }, "PickerSound").start()
    }

    // ══════════════════════════════════════════════════════════════════════
    //  HELPERS
    // ══════════════════════════════════════════════════════════════════════

    /** Clamp a [-1, 1] double to a 16-bit PCM short. */
    private fun toShort(value: Double): Short {
        val clamped = value.coerceIn(-1.0, 1.0)
        return (clamped * Short.MAX_VALUE).toInt().toShort()
    }
}
